llvm/lib/CAS/TreeSchema.cpp

```diff
@@ -0,0 +1,227 @@
+//===- TreeSchema.cpp -----------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/CAS/TreeSchema.h"
+#include "llvm/Support/Endian.h"
+#include "llvm/Support/EndianStream.h"
+#include "llvm/Support/MemoryBufferRef.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/StringSaver.h"
+
+using namespace llvm;
+using namespace llvm::cas;
+
+char TreeSchema::ID = 0;
+constexpr StringLiteral TreeSchema::SchemaName;
+
+void TreeSchema::anchor() {}
+
+bool TreeSchema::isNode(const ObjectHandle &Node) const {
+  // Load the first ref to check its content.
+  if (CAS.getNumRefs(Node) < 1)
+    return false;
+
+  // If can't load the first ref, consume error and return false.
+  auto FirstRef = CAS.readRef(Node, 0);
+  return FirstRef == getKindRef();
+}
+
+TreeSchema::TreeSchema(cas::CASDB &CAS) : TreeSchema::RTTIExtends(CAS) {
+  auto Kind = cantFail(CAS.create(None, SchemaName));
+  TreeKindRef.emplace(Kind.getRef());
+}
+
+ObjectRef TreeSchema::getKindRef() const { return *TreeKindRef; }
+
+size_t TreeSchema::getNumTreeEntries(TreeProxy Tree) const {
+  return CAS.getNumRefs(Tree) - 1;
+}
+
+Error TreeSchema::forEachTreeEntry(
+    TreeProxy Tree,
+    function_ref<Error(const NamedTreeEntry &)> Callback) const {
+  for (size_t I = 0, IE = getNumTreeEntries(Tree); I != IE; ++I)
+    if (Error E = Callback(loadTreeEntry(Tree, I)))
+      return E;
+
+  return Error::success();
+}
+
+Error TreeSchema::walkFileTreeRecursively(
+    CASDB &CAS, const ObjectHandle &Root,
+    function_ref<Error(const NamedTreeEntry &, Optional<TreeProxy>)>
+        Callback) {
+  BumpPtrAllocator Alloc;
+  StringSaver Saver(Alloc);
+  SmallString<128> PathStorage;
```

Comment:
`PathStorage` is shadowed below, and never gets used.  I suggest removing the one inside the loop so that it reuses this one in every iteration, since it can then share an allocation if it ever grows larger than 128 bytes.